rpcserver.go

```diff
@@ -7735,12 +7778,44 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
 			err)
 	}
 
-	// Convert the asset amount into a fixed-point.
-	assetAmount := rfqmath.NewBigIntFixedPoint(req.AssetAmount, 0)
+	invUnits := req.AssetAmount
+
+	// If the invoice was created over a satoshi amount, we need to
+	// calculate the units.
+	if satsMode {
+		invUnits = rfqmath.MilliSatoshiToUnits(
+			amtMsat, *askAssetRate,
+		).ToUint64()
+	}
+
+	// If the invoice is for an asset unit amount smaller than the minimal
+	// transportable amount, we'll return an error, as it wouldn't be
+	// payable by the network.
+	if acceptedQuote.MinTransportableUnits > invUnits {
+		return nil, fmt.Errorf("cannot create invoice over %d asset "+
+			"units, as the minimal transportable amount is %d "+
+			"units with the current rate of %v units/BTC",
+			invUnits, acceptedQuote.MinTransportableUnits,
+			acceptedQuote.AskAssetRate)
+	}
+
+	switch {
+	case satsMode:
+		amtMsat := lnwire.MilliSatoshi(req.InvoiceRequest.ValueMsat +
+			req.InvoiceRequest.Value*(1000))
+
+		iReq.ValueMsat = int64(amtMsat)
+
+	default:
```

Comment:
IMO all this calculation logic with different possible paths has become quite complex and hard to read and hard to test as well. Should we extract the logic into a method and write some unit tests for it?

Reply:
Sounds like a similar sentiment to this comment: https://github.com/lightninglabs/taproot-assets/pull/1448/commits/e5ea7bcf227b9fe6569f618ceec309fa09dcbae1#r2008167666